botocore/http_session.py

```diff
@@ -0,0 +1,216 @@
+        auth_str = '{0}:{1}'.format(username, password)
+        encoded_str = b64encode(auth_str.encode('ascii')).strip().decode()
+        return 'Basic {0}'.format(encoded_str)
+
+    def _get_auth_from_url(self, url):
+        parsed_url = urlparse(url)
+        try:
+            return unquote(parsed_url.username), unquote(parsed_url.password)
+        except (AttributeError, TypeError):
+            return None, None
+
+
+class URLLib3Session(object):
+    """A basic HTTP client that supports connection pooling and proxies.
+
+    This class is inspired by requests.adapters.HTTPAdapter, but has been
+    boiled down to meet the use cases needed by botocore. For the most part
+    this classes matches the functionality of HTTPAdapter in requests v2.7.0
+    (the same as our vendored version). The only major difference of note is
+    that we currently do not support sending chunked requests. While requests
+    v2.7.0 implemented this themselves, later version urllib3 support this
+    directly via a flag to urlopen so enabling it if needed should be trivial.
+    """
+    def __init__(self,
+                 verify=True,
+                 proxies=None,
+                 timeout=None,
+                 max_pool_connections=MAX_POOL_CONNECTIONS,
+    ):
+        self._verify = verify
+        self._proxy_config = ProxyConfiguration(proxies=proxies)
+        if timeout is None:
+            timeout = DEFAULT_TIMEOUT
+        if not isinstance(timeout, (int, float)):
+            timeout = Timeout(connect=timeout[0], read=timeout[1])
+        self._timeout = timeout
+        self._max_pool_connections = max_pool_connections
+        self._proxy_managers = {}
+        self._manager = PoolManager(
+            strict=True,
+            timeout=self._timeout,
+            maxsize=self._max_pool_connections,
+        )
+
+    def _get_proxy_manager(self, proxy_url):
+        if proxy_url not in self._proxy_managers:
+            proxy_headers = self._proxy_config.proxy_headers_for(proxy_url)
+            self._proxy_managers[proxy_url] = proxy_from_url(
+                proxy_url,
+                strict=True,
+                timeout=self._timeout,
+                proxy_headers=proxy_headers,
+                maxsize=self._max_pool_connections
+            )
+
+        return self._proxy_managers[proxy_url]
+
+    def _path_url(self, url):
+        parsed_url = urlparse(url)
+        path = parsed_url.path
+        if not path:
+            path = '/'
+        if parsed_url.query:
+            path = path + '?' + parsed_url.query
+        return path
+
+    def _setup_ssl_cert(self, conn, url, verify):
+        if url.lower().startswith('https') and verify:
+            conn.cert_reqs = 'CERT_REQUIRED'
+            conn.ca_certs = get_cert_path(verify)
+        else:
+            conn.cert_reqs = 'CERT_NONE'
+            conn.ca_certs = None
+
+    def send(self, request):
+        try:
+            request_target = self._path_url(request.url)
```

Comment:
Could you break this method up?  It's fairly lengthy.

I also don't quite follow the logic where `request_target` is set to `self._path_url(request.url)`, but for the proxy case `request_target` is set to just `request.url` (line 170).  Maybe update the comment to explain why we need to use the  full url in the proxy case.

Reply:
Broke it up a little bit more. I've updated the comment to note that proxying an HTTP requires the absolute URI so the host to connect to is part of the request target.